### DISCUSSION TITLE
What code should I look at to see the implementation created by Caffeine newBuilder?

### DISCUSSION THREAD

**Discussion by jojaeng2** [ben-manes/caffeine#1762] [Q&A] [answered]
Hello. The manual provided by Github wiki is provided as follows

``` java
Cache<Key, Graph> cache = Caffeine.newBuilder()
    .expireAfterWrite(10, TimeUnit.MINUTES)
    .maximumSize(10_000)
    .build();

// Lookup an entry, or null if not found
Graph graph = cache.getIfPresent(key);
// Lookup and compute an entry if absent, or null if not computable
graph = cache.get(key, k -> createExpensiveGraph(key));
// Insert or update an entry
cache.put(key, graph);
// Remove an entry
cache.invalidate(key);
```
I'd like to open a Caffeine object created by this newBuilder. I understood that BoundLocalManualCache/UnboundLocalManualCache was created and returned, and that these objects were created using the methodHandle provided by java, but I couldn't find the methodHandle combined with which creator and which Caffeine cache implementation was used.

If you can help me understand the internal behavior, I think it will help me understand. Thank you.

**Comment by ben-manes** [comment, marked as answer]
The implementations are private to the package. The unit tests unwrap to inspect internals, such as 

https://github.com/ben-manes/caffeine/blob/72ec19eb7b15d9c347d62159445d373468f9d762/caffeine/src/test/java/com/github/benmanes/caffeine/cache/BoundedLocalCacheTest.java#L2808-L2810

The build uses a [code generator](https://github.com/ben-manes/caffeine/tree/master/caffeine/src/javaPoet/java/com/github/benmanes/caffeine/cache) for the concrete classes, which the method handle loads. This is to avoid memory waste to minimize field allocations, as only a few types might be used in a given application. The generated code is part of the sources jar (so your IDE should find it) or can be created by running `grades compileJava`.

**Reply by jojaeng2** [reply]
Thank you for your answer, sir
Actually that wasn't the intention of my question, but it seems to have been solved in the process of opening more codes in IDE. Sorry for not explaining it properly.
